lib/compilation_engine.rb

```diff
@@ -231,7 +231,9 @@ def compile_expression_list
 
   def compile_do
     advance # do
-    compile_subroutine_call
+    @expression_parser.parse_term.write_vm_code(@vm_writer)
+    advance # ;
+    @vm_writer.write_pop(:TEMP, 0)
```

Comment:
Yeah, this is much better. It was kind of weird that `#compile_subroutine_call` was responsible for advancing past the `;` token (because that’s really part of the `do` syntax, not the subroutine call itself), so I’m glad it’s gone away. 🎉